fun/dirs: Fix $HOME selection in mcd and cd with a number

pghvlaans opened the pull request:
In `fun/dirs`, the `mcd` function is broken:

* When selecting a number, `\cd` is called with that number, resulting in failed or (very occasionally) incorrect directory changes.
* When selecting a directory under `$HOME` by name, `\cd` attempts to change directory to a directory starting with escaped `~`. The results are similar.

In addition, `cd` no longer interacts with the push stack. In the original state of `fun/dirs` (added to the repository in early 2001, but previously appearing in Bolsky and Korn), using `cd 5` would change to the directory in the fifth position on the push stack, for example. 

`fun/dirs`: In the function `_cd`, clear all positional parameters when `mcd` is called and a directory under `$HOME` is selected by name, and whenever `cd` is called with a number as the first parameter. This allows `\cd` to use `$dirs` as intended in such cases.

pghvlaans committed:
fun/dirs: Fix numeric selection in mcd.

pghvlaans committed:
fun/dirs: Fix names with mcd.

pghvlaans committed:
fun/dirs: Reset _mcd_called.

pghvlaans commented:
As a side note, these commits make it so that e.g. `cd 5` or `cd -2` attempt to change directory to literal `5` and `-2`, respectively, rather than whatever is in position on the push stack.

I couldn't find any documentation from the time `dirs` was introduced (the first half of 2001) to indicate whether those kinds of `cd` calls should work, but enabling that behavior would be a simple change. Please let me know which would be preferable.

pghvlaans committed:
fun/dirs: cd with a number uses the put stack.

pghvlaans committed:
fun/dirs: cd with initial number overrides other arguments

pghvlaans commented:
After a little more testing, it turned out that in the original `fun/dirs`, something like `cd 2 whatever` would always execute `cd 2`.

pghvlaans committed:
Merge branch 'ksh93:dev' into dirs

McDutchie commented on src/cmd/ksh93/fun/dirs:
```diff
@@ -30,16 +32,28 @@ function _cd
+        while (($#))
+        do
+              shift
+        done
```
The positional parameters can be emptied all at once using `set --`, which is more efficient than removing them one by one.

pghvlaans replied:
That makes sense. Thank you.

McDutchie commented on src/cmd/ksh93/fun/dirs:
```diff
@@ -55,6 +69,13 @@ function _cd
     esac
+    # mcd passes escaped ~, which must be dealt with
+    if     ((_mcd_called>0))
+    then   case "$@" in
```
Using `"$@"` in this context makes no sense as that quoted expansion generates one word per positional parameter, but only exactly one word is possible between `case` and `in`, so the behaviour of `"$@"` is unspecified in this context. On ksh, in practice it's the same as `"$*"` here, i.e., the positional parameters get joined into one string separated by the first character of `$IFS`. But in this case you're only checking the first character of the first parameter, so you might as well just use `$1` (no quotes needed).

pghvlaans replied:
Yes, only the first parameter would be wanted here.

pghvlaans committed:
Tweak for review comments

McDutchie committed:
more tweaks

- Make PS3 a local variable so global PS3 is not overwritten.
- 'return' instead of 'return 1' when \cd fails so that its actual
  exit status is passed down (as of abf87cb7, the exit status of cd
  may be 2).
- Delete ancient code for TERM=630. Presumably this is the AT&T 630
  terminal from the 1980s:
  https://terminals-wiki.org/wiki/index.php/AT%26T_630_MTG
  On modern systems the corresponding TERM value would be att630
  (as per my /usr/share/terminfo directory).
  The code writes the PWD to the label area, as on page 5-23 here:
  http://bitsavers.org/pdf/att/630_mtg/630_MTG_Terminal_Users_Guide_Jul87.pdf
  It's interesting, but dead code for ±100% of current users.
- mcd(): Quote "$REPLY" when calling _cd to avoid problems with
  whitespace, field splitting and globbing.
